src/test/ui/consts/const-loop.rs

```diff
@@ -1,13 +1,42 @@
-const _: i32 = {
-    let mut x = 0;
+const _: () = loop {}; //~ ERROR loop is not allowed in a const
+
+static FOO: i32 = loop { break 4; }; //~ ERROR loop is not allowed in a static
+
+const fn foo() {
+    loop {} //~ ERROR loop is not allowed in a const fn
+}
+
+pub trait Foo {
+    const BAR: i32 = loop { break 4; }; //~ ERROR loop is not allowed in a const
+}
 
-    while x < 4 {
+impl Foo for () {
+    const BAR: i32 = loop { break 4; }; //~ ERROR loop is not allowed in a const
+}
+
+fn nested() {
+    const fn fns() {
+        loop {} //~ ERROR loop is not allowed in a const fn
+    }
+}
+
+fn main() {
+    let x = [0; {
+        while false {}
         //~^ ERROR constant contains unimplemented expression type
         //~| ERROR constant contains unimplemented expression type
+        4
```

Comment:
Notably, the HIR checker doesn't process array lengths. I'm not sure how to do this. I think it may need a different `NestedVisitorMap`? We can't turn anything in the MIR checker into a `delay_span_bug` until this is handled. Also, we rely on the fact that MIRI will execute an array length that fails to const-check in some tests, so we would have to do this another way.

Reply:
Maybe https://doc.rust-lang.org/nightly/nightly-rustc/rustc_passes/loops/struct.CheckLoopVisitor.html#method.visit_anon_const can provide some inspiration?

Reply:
`walk_anon_const` calls `visit_nested_body`, so it's something to do with the `NestedVisitorMap`. I'll try the more expensive version in a bit, but I would consider putting that in its own PR, since it will require non-trivial changes to 4-5 tests.

Reply:
Curious; Worth noting that `CheckLoopVisitor` also uses `OnlyBodies`. That said, if you want to keep this PR more limited and follow up with more changes to it I'm happy with that. Up to you. :)

Reply:
Hmm, yeah maybe I'm mistaken or there's something else going on.

Reply:
I followed `CheckLoopVisitor`s example, so I believe things are working correctly now. I didn't actually have to change the tests, since MIRI *still* runs, even if errors occur in this pass.